### PR TITLE
`spirv-remap`: Support outputting each SPIR-V module to a filename

### LINKED ISSUE #3215: `spirv-remap`: Output to single directory or one or more files
The `spirv-remap` tool currently outputs shaders to a single directory.

This can complicate integration with build systems that explicitly track both inputs and outputs produced by an executable.

The `spirv-remap` tool should support:
  * Outputting one or more inputs to a single directory -- the current behavior.
  * One output file per input -- new behavior.

This new behavior can be robustified using std::filesystem::is_directory().

### PR DESCRIPTION
**Pull request by allanmac**
Closes #3215.
    
This PR enables:
* Outputting one or more inputs to a single directory (current behavior).
* One output file per input (new behavior).
    
If there is only a single output then the output is probed to determine if it's a directory using C++17 `std::filesystem::is_directory()`.

### CODE REVIEW

**Comment by allanmac** [comment]
Why do Bazel builds use `"-std=c++11"` instead of C++17?

**Comment by arcady-lunarg** [comment]
I'm going to take a look at this today.

**Comment by arcady-lunarg** [comment]
Actually, rather than adding the ifdefs for c++17, I'd like to switch everything to using c++17 first. I opened #3230 for that, once that is merged you can rebase your change on it.

**Review by arcady-lunarg — CHANGES_REQUESTED** [review]
@allanmac given that #3230 has been merged, could you rework this to remove the ifdefs for C++17? That's the officially required minimum version of C++ per the README anyway.

**Comment by allanmac** [comment]
Yes, I will rework now that C++17 is universal.

**Comment by allanmac** [comment]
If I'm reading it correctly, MSVC has supported C++17 (and `<filesystem>`) since VS 2017 15.7 using the `/std:c++17` switch.

Is this another example of a missing C++17 switch? 

---

This is the error I'm seeing in CI:

`CI-windows-msvc-2019-release-bazel — CI build failed`:

```
C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.34.31933\include\filesystem(13): warning STL4038: The contents of <filesystem> are available only with C++17 or later.
StandAlone/spirv-remap.cpp(394): error C2039: 'filesystem': is not a member of 'std'
C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.34.31933\include\algorithm(116): note: see declaration of 'std'
StandAlone/spirv-remap.cpp(394): error C3083: 'filesystem': the symbol to the left of a '::' must be a type
StandAlone/spirv-remap.cpp(394): error C2039: 'is_directory': is not a member of 'std'
C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.34.31933\include\algorithm(116): note: see declaration of 'std'
StandAlone/spirv-remap.cpp(394): error C3861: 'is_directory': identifier not found
```

**Comment by arcady-lunarg** [comment]
I think I missed some compiler flags for windows in the bazel build. I'll try to add those flags to force c++17 mode in that case.

**Review by allanmac — COMMENTED** [review]
Removed C++2017 dependency.

**Comment by arcady-lunarg** [comment]
Sorry this has been taking so long to get merged, it turns out our android CI build was not up to standards and needs to be reworked a bit, so that C++17 works there as well. However, that is now on my todo list, and once that's in, we can rebase this PR and merge it.

File: StandAlone/spirv-remap.cpp [hunk 1]
```diff
@@ -361,13 +384,27 @@ int main(int argc, char** argv)
...
+        usage(argv[0], "Output directory or file(s) required.");
 
-    if (outputDir.empty())
-        usage(argv[0], "Output directory required");
+    const bool isMultiInput      = inputFiles.size() > 1;
+    const bool isMultiOutput     = outputDirOrFiles.size() > 1;
+    const bool isSingleOutputDir = !isMultiOutput && std::filesystem::is_directory(outputDirOrFiles[0]);
+
+    if (isMultiInput && !isMultiOutput && !isSingleOutputDir)
+        {
```
Comment (arcady-lunarg) [inline StandAlone/spirv-remap.cpp:397]: In this case, I would omit the braces entirely.

**Review by arcady-lunarg — CHANGES_REQUESTED** [review]
The basic design of this is solid, though I haven't tested it yet. However, could you please make sure your new code follows the formatting conventions of the existing code?

File: StandAlone/spirv-remap.cpp [hunk 2]
```diff
@@ -182,32 +186,49 @@ namespace {
...
-            const std::string &filename = *it;
-            std::vector<SpvWord> spv;
-            read(spv, filename, verbosity);
-            spv::spirvbin_t(verbosity).remap(spv, whiteListStrings, opts);
-            const std::string outfile = outputDir + path_sep_char() + basename(filename);
-            write(spv, outfile, verbosity);
-        }
+        for (std::size_t ii=0; ii<inputFiles.size(); ii++)
+            {
+                std::vector<SpvWord> spv;
```
[the remark this answers is not in the document]
Comment (allanmac) [reply StandAlone/spirv-remap.cpp]: Done.

File: StandAlone/spirv-remap.cpp [hunk 3]
```diff
@@ -172,7 +176,7 @@ namespace {
             << " [--strip-all | --strip all | -s]"
             << " [--strip-white-list]"
             << " [--do-everything]"
-            << " --input | -i file1 [file2...] --output|-o DESTDIR"
+            << " --input | -i file1 [file2...] --output | -o destdir | destfile1 [destfile2...]"
```
[the remark this answers is not in the document]
Comment (allanmac) [reply StandAlone/spirv-remap.cpp]: ACK

File: StandAlone/spirv-remap.cpp [hunk 4]
```diff
@@ -182,32 +186,49 @@ namespace {
...
+
+                spv::spirvbin_t(verbosity).remap(spv, whiteListStrings, opts);
+
+                if (isSingleOutputDir)
+                    {
+                        // write all outputs to same directory
+                        const std::string outFile = outputDirOrFiles[0] + path_sep_char() + basename(inputFiles[ii]);
+                        write(spv, outFile, verbosity);
+                    }
+                else
```
[the remark this answers is not in the document]
Comment (allanmac) [reply StandAlone/spirv-remap.cpp]: Whoops, done.

File: StandAlone/spirv-remap.cpp (hunk 1 above)
Comment (allanmac) [reply StandAlone/spirv-remap.cpp]: ACK/Done.

**Review by arcady-lunarg — APPROVED** [review]
LGTM

**Pull request merged**